### DISCUSSION TITLE
Variance in input lag makes rhythm games unplayable

### DISCUSSION THREAD

**Discussion by taboodles** [ndeadly/MissionControl#292] [General]
No hard feelings or anything, I am very impressed with this project. I just wanted to add a data point.

I've been using sys-con to use a Dualshock 4 controller to play _Hatsune Miku: Project DIVA Mega Mix_ for almost a year now, and haven't had problems. The game has lets you calibrate lag, so any delay caused by the display or controllers can be mitigated if the delay is stable.

Even with several attempts at calibration, I can't reliably hit notes using the Dualshock 4 and MissionControl. Just based on my sense of rhythm, it seems like all controller inputs eventually register, but 20% of them get noticeably delayed.

Based on some googling, arcade rhythm games have a leniency of at most 50 milliseconds to get the best rating for each note. I'd estimate that to be an upper bound for how much jitter can be present for rhythm games to be playable.

**Comment by ndeadly** [comment]
Do you have `nx-btred` installed by any chance? Another user reported pretty much exactly the same issue as you and found that to be the cause of the latency https://github.com/ndeadly/MissionControl/issues/246

**Reply by taboodles** [reply]
I wasn't using `nx-btred`, no. Reading that issue and seeing that the reporter had success with Project Diva did encourage me to try again.

I've found that if my Dualshock4 is my only connected controller with the Switch, I have no gameplay issues whatsoever. I can even connect two Dualshock4 controllers and still play smoothly. I experimented a bit and I'm able to reliably make things bad again by connecting my Switch Pro Controller, which had always been connected when I first noticed problems.

I guess the presence of a synced official Nintendo Switch Pro Controller causes the issue? This makes sense to me based on how you described MissionControl's mitm business.

**Reply by ndeadly** [reply]
That's quite strange, I'd expect connecting a second DS4 to be more likely to create problems since they send input data at a higher rate. Which firmware are you using? Does the order in which you connect the controllers make any difference?

**Reply by taboodles** [reply]
I'm on 12.0.2 with atmosphere 19.4. My current homebrew setup comes `sdsetup.com`'s "Recommended Defaults". In addition, I have sys-con, MissionControl, and a set of sigpatches.

I see stuttering input whether or not the Dualshock4 is connected first or second in comparison to the SwitchProController. When they are both connected, the official Switch controller works fine and I am able to play normally.

**Reply by ndeadly** [reply]
Here's a [build](https://github.com/ndeadly/MissionControl/files/6770746/MissionControl-0.5.0-alpha.4-ds4-polling-rate-1b09ede.zip) I made that attempts to reduce the default reporting rate of the DS4 controller to the same as that of a Switch Pro controller (1000Hz -> 125Hz). Try it and see if it makes any difference

**Reply by taboodles** [reply]
I don't notice any difference using the new build,
Dualshock4 alone is still good,
Dualshock4 and Dualshock4 is still good,
Dualshock4 and Switch Pro Controller is still bad.

Edit: Although maybe the tweak improved the situation compared to earlier, and I'm just unable to detect it. My only sense of whether there is input stutter is by playing a song and seeing if I can sustain a note combo.

**Reply by ndeadly** [reply]
It may have made a small difference if anything was being overloaded by the high data rate, but is obviously not the root cause. It's super weird that it looks to somehow have something to do with the official pro controller being connected. Can you check whether the same thing happens with joycons connected wirelessly too? I wonder whether it has anything to do with this time slot index thing I stubbed out for unofficial controllers... I will post a build for you to try later to test the theory

**Reply by taboodles** [reply]
I'm still using your hand-crafted build.
I'm seeing the same input jitter with my Dualshock4 inputs when Dual Joycons are connected. The Dual Joycons' inputs seem to have no delay problems.

**Reply by ndeadly** [reply]
See if this does anything for you

[MissionControl-0.5.0-alpha.4-no-tsi-stub-67deb90.zip](https://github.com/ndeadly/MissionControl/files/6790602/MissionControl-0.5.0-alpha.4-no-tsi-stub-67deb90.zip)

**Reply by taboodles** [reply]
Hey! That seemed to totally do the trick! I don't have any gameplay issues at all, and everything still seems stable. I did gameplay tests with the Switch Pro Controller, 2 Dualshock controllers, and my Joycons all connected simultaneously.

**Reply by ndeadly** [reply]
Cool! Wasn't really expecting that to work so that's interesting. This also didn't include the reporting rate limiting so seems it's purely because of this function. I guess I'll have to work on only blocking this command on controllers where it causes problems. Thanks for testing.

**Reply by ndeadly** [reply]
Also, FYI that build above in its current form will break connectivity for Xbox One and Gamesir controllers as they can't handle that vendor command (that was the reason I blocked it in the first place)

**Reply by ndeadly** [reply]
This fix and the controller report rate reduction are both included in the new beta release I made just now. Hopefully that's the end of your input jitter :)

https://github.com/ndeadly/MissionControl/releases/tag/v0.5.0-beta

**Comment by Masamune3210** [comment]
I would rebuild your card manually without using any pre-packaged setups and test again, they tend to cause more issue than good as it's entirely in the air as to what is installed as their "defaults"
